### PR TITLE
🚀 APE-146: Refactor currency validation to handle various formats and symbols

### PR DESCRIPTION
**Pull request by mohamedmamdouh22**
1. Handle more currency cases like 
  - ("€1.234", 1234.0)
  - ("Starting at: 12:00", 12.0)
  - ("Starting At 12.99", 12.99)
  - ("From 399.99", 399.99),
2. Add more unit tests

### CODE REVIEW

**Comment by awtkns** [comment]
Screenshot 2024-08-27 at 4 16 50 PM
Would recommend either deleting or creating a new branch for each pr so that the commit history doesnt include old stuff

**Comment by awtkns** [comment]
to add to this, ideally call the branch APE-<ticket number> then it will get linked properly. :)

File: harambe/parser/type_currency.py [hunk 1]
```diff
@@ -10,14 +10,31 @@ def __new__(cls) -> Any:
...
+        if cleaned_value.startswith("."):  # fraction
+            cleaned_value = "0" + cleaned_value
+            return float(cleaned_value)
+        if "," in cleaned_value and "." in cleaned_value:
+            if cleaned_value.index(",") < cleaned_value.index("."):
+                cleaned_value = cleaned_value.replace(",", "")
+            else:
+                cleaned_value = cleaned_value.replace(".", "").replace(",", ".")
+        elif "," in cleaned_value and "." not in cleaned_value:
+            cleaned_value = cleaned_value.replace(",", "")
```
Comment (awtkns) [inline harambe/parser/type_currency.py:35]: This is a big blob of code. lets split it up for readability

```suggestion
        if ":" in value:  # cases like 12:00
            value = re.sub(r"(?<=\d):(?=\d)", ".", value)
            
        cleaned_value = re.sub(r"[^\d.,-]", "", value)
        cleaned_value = re.sub(r"^0+(?!$)", "", cleaned_value)
        
        if "." in cleaned_value:
            decimal_parts = cleaned_value.split(".")
            if len(decimal_parts[-1]) == 3:  # thousands separators
                cleaned_value = cleaned_value.replace(".", "")
                
        if cleaned_value.startswith("."):  # fraction
            cleaned_value = "0" + cleaned_value
            return float(cleaned_value)
            
        if "," in cleaned_value and "." in cleaned_value:
            if cleaned_value.index(",") < cleaned_value.index("."):
                cleaned_value = cleaned_value.replace(",", "")
            else:
                cleaned_value = cleaned_value.replace(".", "").replace(",", ".")
        elif "," in cleaned_value and "." not in cleaned_value:
            cleaned_value = cleaned_value.replace(",", "")
```

Comment (mohamedmamdouh22) [reply harambe/parser/type_currency.py:35]: yeah you are right, it is getting cloudy

File: test/parser/test_type_currency.py [hunk 2]
```diff
@@ -41,6 +50,10 @@ def test_flexible_currency_success(input_value, expected_output):
         "one hundred",
         "1.00.00",
         "1.2.3",
+        "Begin 12.00 End 23.00",
+        "Between 12.00 And 23.00",
+        "From 12.00 To 23.00",
+        "12.00 - 23.00",
         "",
```
Comment (awtkns) [inline test/parser/test_type_currency.py:57]: some more:
```
 "Not a number",
    "1.234.56",  # Ambiguous separators
     "1,234,56",  # Ambiguous separators
    "$",  # Just a currency symbol
    "1,234.56.78",  # Multiple decimal points
```

File: test/parser/test_type_currency.py [hunk 3]
```diff
@@ -25,8 +25,17 @@ class TestModel(BaseModel):
...
+        ("1.234€", 1234.0),
+        ("€1.234", 1234.0),
         ("1,234$", 1234.0),
         ("1,234 $", 1234.0),
+        ("1.234,45", 1234.45),
+        ("1.234.456,00", 1234456.00),
+        ("1.000.000", 1000000.00),
+        ("Starting at: 12:00", 12.0),
+        ("Starting At 12.99", 12.99),
+        ("From 399.99", 399.99),
```
Comment (awtkns) [inline test/parser/test_type_currency.py:38]: can we add a fraction case eg: 0.0004 $
this is valid for stuff like crypto or stocks

Comment (mohamedmamdouh22) [reply test/parser/test_type_currency.py:38]: done!

File: harambe/parser/type_currency.py [hunk 4]
```diff
@@ -10,14 +10,35 @@ def __new__(cls) -> Any:
...
     @staticmethod
     def validate_currency(value: str) -> float:
-        if isinstance(value, float):
-            return value
+        if isinstance(value, (float, int)):
+            return float(value)
 
         value = str(value).strip()
+        cleaned_value = re.sub(r"[^\d.,-]", "", value)
+        cleaned_value = re.sub(r"^0+(?!$)", "", cleaned_value)
```
Comment (awtkns) [inline harambe/parser/type_currency.py:18]: nit: in the future we can pre compile this regex as an optimization
```python
   CLEAN_VALUE_PATTERN = re.compile(r"[^\d.,-]")
    LEADING_ZEROS_PATTERN = re.compile(r"^0+(?!$)")

    @staticmethod
    def validate_currency(value: Union[str, float, int]) -> float:
        if isinstance(value, (float, int)):
            return float(value)
        value = str(value).strip()
        cleaned_value = CurrencyValidator.CLEAN_VALUE_PATTERN.sub("", value)
        cleaned_value = CurrencyValidator.LEADING_ZEROS_PATTERN.sub("", cleaned_value)
```

**Review by awtkns — APPROVED** [review]

**Pull request merged**